framework-plugins/lisk-framework-forger-plugin/src/forger_plugin.ts

```diff
@@ -288,98 +289,116 @@ export class ForgerPlugin extends BasePlugin {
 			});
+			await setForgerInfo(this._forgerPluginDB, forgerAddressBinary, { ...forgerInfo });
 		}
 
-		for (const trx of payload) {
-			if (trx.type === VoteTransaction.TYPE) {
-				for (const vote of (trx.asset as Asset).votes) {
-					const registeredDelegateIndex = this._forgersList.findIndex(
-						forger => forger.address === vote.delegateAddress,
-					);
-					if (registeredDelegateIndex !== -1) {
-						const delegateVoteIndex = forgerInfo.votesReceived.findIndex(aVote =>
-							aVote.address.equals(Buffer.from(vote.delegateAddress, 'base64')),
-						);
-						if (delegateVoteIndex === -1) {
-							forgerInfo.votesReceived.push({
-								address: Buffer.from(vote.delegateAddress, 'base64'),
-								amount: BigInt(vote.amount),
-							});
-						} else {
-							forgerInfo.votesReceived[delegateVoteIndex].amount += BigInt(vote.amount);
-						}
-						isUpdated = true;
-					}
-				}
-			}
-		}
-
+		await this._addVotesReceived(payload);
 		await this._updateMissedBlock(encodedBlock);
-
-		if (isUpdated) {
-			await setForgerInfo(this._forgerPluginDB, forgerAddressBinary, { ...forgerInfo });
-		}
 	}
 
-	private async _decrementForgerInfo(encodedBlock: string): Promise<void> {
+	private async _revertForgerInfo(encodedBlock: string): Promise<void> {
 		const {
 			forgerAddress,
 			forgerAddressBinary,
 			header: { reward },
 			payload,
 		} = this._getForgerHeaderAndPayloadInfo(encodedBlock);
 		const forgerInfo = await getForgerInfo(this._forgerPluginDB, forgerAddressBinary);
-		let isUpdated = false;
 
 		if (this._forgersList.find(forger => forger.address === forgerAddress)) {
 			forgerInfo.totalProducedBlocks -= 1;
 			forgerInfo.totalReceivedRewards -= BigInt(reward);
 			forgerInfo.totalReceivedFees -= this._getFee(payload, encodedBlock);
-			isUpdated = true;
+			await setForgerInfo(this._forgerPluginDB, forgerAddressBinary, { ...forgerInfo });
 		}
 
+		await this._revertVotesReceived(payload);
+	}
+
+	private async _addVotesReceived(payload: ReadonlyArray<TransactionJSON>): Promise<void> {
+		const batch = this._forgerPluginDB.batch();
 		for (const trx of payload) {
 			if (trx.type === VoteTransaction.TYPE) {
+				const senderAddress = getAddressFromPublicKey(Buffer.from(trx.senderPublicKey, 'base64'));
 				for (const vote of (trx.asset as Asset).votes) {
-					const delegateVoteIndex = forgerInfo.votesReceived.findIndex(aVote =>
-						aVote.address.equals(Buffer.from(vote.delegateAddress, 'base64')),
+					const registeredDelegateIndex = this._forgersList.findIndex(
+						forger => forger.address === vote.delegateAddress,
 					);
-					if (delegateVoteIndex !== -1) {
-						forgerInfo.votesReceived[delegateVoteIndex].amount -= BigInt(vote.amount);
-						if (forgerInfo.votesReceived[delegateVoteIndex].amount === BigInt(0)) {
-							forgerInfo.votesReceived.splice(delegateVoteIndex, 1);
+					if (registeredDelegateIndex !== -1) {
+						const affectedForgerInfo = await getForgerInfo(
```

Comment:
This should be done outside of loop?
Also, if there are multiple transaction from the same sender in the same block
- if first transaction votes, and second transaction unvotes, it will have invalid result
- multiple get runs which is bit inefficient